### PR TITLE
cpu/cc2538: gpio: save a few cycles in handle_isr()

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

If only one bit it is set in `state` (one GPIO pin caused an interrupt), don't loop over all 8 bits.

Use `clz` to get the position of the first interrupt bit and clear it, looping only as many times as there are actual interrupts.

### Testing procedure

<details>
<summary>All 8 interrupt pins/port should still work.</summary>

```C
#include <stdio.h>
#include <stdint.h>

static void handle_isr(uint32_t state) {
    while (state) {
        int pin = 8 * sizeof(state) - __builtin_clz(state) - 1;
        state &= ~(1 << pin);
        printf("Interrupt on pin %d\n", pin);
    }
}

int main(int argc, char **argv) {
    uint32_t state = 1;

    for (int i = 0; i < 8; ++i) {
        printf("%d:\n", i);
        handle_isr(state);

        state <<= 1;
        if (state & 0x4) {
            state |= 1;
        }
    }

}
```
</details>

I tested this on the `openmote-b` and both the user button and the interrupts generated by the at86rf215 radio would still work.

### CODE REVIEW

File: cpu/cc2538/periph/gpio.c [hunk 1]
```diff
@@ -213,10 +213,11 @@ static inline void handle_isr(uint8_t port_num)
     port->IC             = 0x000000ff;
     port->IRQ_DETECT_ACK = (0xff << (port_num * GPIO_BITS_PER_PORT));
 
-    for (int i = 0; i < GPIO_BITS_PER_PORT; i++) {
-        if (state & (1 << i)) {
-            isr_ctx[port_num][i].cb(isr_ctx[port_num][i].arg);
-        }
+    state &= GPIO_BIT_MASK;
+    while (state) {
```
Comment (kaspar030) [inline cpu/cc2538/periph/gpio.c]: if using builtins, wouldn't sth like:

```
int pin;
while ((pin = __builtin_ffs(state))) {
  isr_ctx[port_num][pin].cb(isr_ctx[port_num][pin].arg);
  state &= ~(1 << pin);
}
```

... save some arithetic?

Comment (benpicco) [reply cpu/cc2538/periph/gpio.c]: The Cortex-M3 only has `clz`, no `ffs`
http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.100165_0201_00_en/ric1414056333562.html

Comment (kaspar030) [reply cpu/cc2538/periph/gpio.c]: but gcc's __builtin_ffs() is implemented using clz?

Comment (benpicco) [reply cpu/cc2538/periph/gpio.c]: Want to play a [game of golf](https://godbolt.org/z/Lfo23M)? 😉

Comment (kaspar030) [reply cpu/cc2538/periph/gpio.c]: Godbolt is awesome! Thanks! :wink:

File: cpu/cc2538/periph/gpio.c [hunk 2]
```diff
@@ -213,10 +213,11 @@ static inline void handle_isr(uint8_t port_num)
     port->IC             = 0x000000ff;
     port->IRQ_DETECT_ACK = (0xff << (port_num * GPIO_BITS_PER_PORT));
 
-    for (int i = 0; i < GPIO_BITS_PER_PORT; i++) {
-        if (state & (1 << i)) {
-            isr_ctx[port_num][i].cb(isr_ctx[port_num][i].arg);
-        }
+    state &= (1 << GPIO_BITS_PER_PORT) - 1;
```
Comment (fjmolinas) [inline cpu/cc2538/periph/gpio.c]: Is this masking all non `gpio` realted bits in `MIS`? If so add a comment.

File: cpu/cc2538/periph/gpio.c [hunk 3]
```diff
@@ -213,10 +213,11 @@ static inline void handle_isr(uint8_t port_num)
...
+    while (state) {
+        int pin = 8 * sizeof(state) - __builtin_clz(state) - 1;
```
Comment (fjmolinas) [inline cpu/cc2538/periph/gpio.c]: Is the 8 for 8 bit per `uint8_t`, is there a macro for this?

Comment (benpicco) [reply cpu/cc2538/periph/gpio.c]: Yea sorry that was silly.
spelling out the 32 bit from which the number of leading zeros is subtracted makes it much clearer that the result must be the position most significant set bit.

**Pull request merged**